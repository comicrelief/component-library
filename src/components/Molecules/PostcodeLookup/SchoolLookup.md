# SchoolLookup

```js
import SchoolLookup from './SchoolLookup';

// This is just an example of how a parent component might handle fetch errors within the component.
const [enterManually, setEnterManually] = React.useState(false);

const fetchErrorHandler = () => {
  setEnterManually(true);
}

enterManually
  ? 'Sorry, there appears to be a problem. Please enter the school\'s details manually.' 
  : (
    <SchoolLookup onSelect={school => alert(JSON.stringify(school, null, 2))} fetchErrorHandler={fetchErrorHandler} />
  )
```
