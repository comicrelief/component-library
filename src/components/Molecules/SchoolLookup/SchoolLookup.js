import React, { useState } from 'react';
import PropTypes from 'prop-types';
import axios from 'axios';

import Input from '../../Atoms/Input/Input';
import {
  Container,
  Dropdown,
  DropdownList,
  DropdownItem,
  DropdownItemSelectable
} from './SchoolLookup.style';
import Text from '../../Atoms/Text/Text';

const PLACEHOLDER = 'Type to start search';
const INPUT_LABEL = 'Enter the name or postcode of your school or nursery';
const DROPDOWN_TIP = 'Please select a school from the list below';
const DEFAULT_LOOKUP_ERROR_MESSAGE =
  'Sorry, there was an unexpected error. Please try again';

const lookup = async (query, config = {}) => {
  const response = await axios.get(
    'https://lookups.sls.comicrelief.com/schools/lookup',
    { params: { query }, ...config }
  );
  return response.data.data.schools;
};

/**
 * Todo:
 *  - Proper styling
 *  - Responsiveness
 *  - Accessibility
 *  - Do we need to process / sanitize / url encode the query in any way?
 *  - Do we need to send any errors to Sentry or any other service?
 *  - Tests
 *  - ??
 */
const SchoolLookup = ({ onSelect, notFoundMessage, lookupErrorMessage }) => {
  const initialState = {
    query: '',
    schoolOptions: [],
    errorMsg: '',
    timeoutId: 0,
    cancelTokenSource: null
  };
  const [state, setState] = useState(initialState);
  const updateState = updatedValues =>
    setState(prevState => ({ ...prevState, ...updatedValues }));

  const lookupHandler = async e => {
    const query = e.currentTarget.value;
    let { errorMsg, schoolOptions } = initialState;
    // clearing error and dropdown as soon as input changes (seemed to me to be the nicest UX.)
    updateState({ query, errorMsg, schoolOptions });

    clearTimeout(state.timeoutId);
    // cancel previous axios request if it was already initiated (before we attempted to clear the timeout above.)
    if (state.cancelTokenSource) {
      state.cancelTokenSource.cancel();
    }

    const timeoutId = setTimeout(async () => {
      let cancelled = false;
      const trimmed = query.trim();

      if (trimmed.length > 2) {
        try {
          const source = axios.CancelToken.source();
          updateState({ cancelTokenSource: source });
          schoolOptions = await lookup(trimmed, { cancelToken: source.token });
          if (schoolOptions.length === 0) {
            errorMsg = notFoundMessage;
          }
        } catch (err) {
          if (axios.isCancel(err)) {
            cancelled = true;
          } else if (lookupErrorMessage) {
            errorMsg = lookupErrorMessage;
          } else {
            // In case we want to handle any errors further up the tree.
            // (need to provide an empty string to the lookupErrorMessage prop to get here).
            throw err;
          }
        }

        if (cancelled === false) {
          updateState({ schoolOptions, errorMsg });
        }
      }
    }, 50);

    updateState({ timeoutId });
  };

  const getSelectHandler = school => () => {
    setState(initialState);
    onSelect(school);
  };

  return (
    <Container>
      <SchoolLookupInput
        value={state.query}
        onChange={lookupHandler}
        errorMsg={state.errorMsg}
      />
      {state.schoolOptions.length > 0 && (
        <Dropdown>
          <DropdownList>
            <DropdownItem>
              <Text>{DROPDOWN_TIP}</Text>
            </DropdownItem>
            {state.schoolOptions.map(school => (
              <DropdownItemSelectable
                key={school.id}
                onClick={getSelectHandler(school)}
              >
                <Text>{`${school.name}, ${school.post_code}`}</Text>
              </DropdownItemSelectable>
            ))}
          </DropdownList>
        </Dropdown>
      )}
    </Container>
  );
};

const SchoolLookupInput = props => (
  <Input
    {...{
      ...props,
      id: 'school-lookup-input',
      type: 'text',
      name: 'query',
      label: INPUT_LABEL,
      placeholder: PLACEHOLDER
    }}
  />
);

SchoolLookup.propTypes = {
  onSelect: PropTypes.func.isRequired,
  notFoundMessage: PropTypes.string.isRequired,
  lookupErrorMessage: PropTypes.string
};

SchoolLookup.defaultProps = {
  lookupErrorMessage: DEFAULT_LOOKUP_ERROR_MESSAGE
};

export default SchoolLookup;
