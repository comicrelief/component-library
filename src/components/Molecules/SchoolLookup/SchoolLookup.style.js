import styled from 'styled-components';

const Container = styled.div`
  position: relative;
`;

const Dropdown = styled.div`
  position: absolute;
  left: 0;
  max-height: 300px;
  max-width: 500px;
  overflow: auto;
  background-color: white;
  border: 1px solid lightgrey;
  width: 100%;
`;

const DropdownList = styled.ul`
  list-style: none;
  padding: 0;
  margin: 0;
`;

const DropdownItem = styled.li`
  line-height: 1.5rem;
  padding: 0.5rem;
`;

const DropdownItemSelectable = styled(DropdownItem)`
  cursor: pointer;
  border-top: 1px solid lightgrey;

  &:hover {
    background-color: lightgrey;
  }
`;

export {
  Container,
  Dropdown,
  DropdownList,
  DropdownItem,
  DropdownItemSelectable
};
