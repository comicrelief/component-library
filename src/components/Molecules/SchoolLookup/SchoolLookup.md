# SchoolLookup

```js
import SchoolLookup from './SchoolLookup';

// This error count stuff is just an example of how your parent
//  might handle repeated errors (in case the school lookup api stops working)
const [errorCount, setErrorCount] = React.useState(0);

const fetchErrorHandler = (error, errorCount) => {
  setErrorCount(prevCount => prevCount + 1);

  return 'Sorry, there was an unexpected error. Please try again';
}

errorCount > 1 
  ? 'Sorry, there appears to be a problem. Please enter the school\'s details manually.' 
  : (
    <SchoolLookup onSelect={school => alert(JSON.stringify(school, null, 2))} fetchErrorHandler={fetchErrorHandler} />
  )
```
