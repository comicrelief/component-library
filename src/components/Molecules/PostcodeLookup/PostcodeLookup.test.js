import React from 'react';
import 'jest-styled-components';

import renderWithTheme from '../../../hoc/shallowWithTheme';
import SchoolLookup from './SchoolLookup';

it('renders correctly', () => {
  const renderer = renderWithTheme(<SchoolLookup onSelect={() => {}} />);

  expect(renderer.toJSON()).toMatchSnapshot();
});
