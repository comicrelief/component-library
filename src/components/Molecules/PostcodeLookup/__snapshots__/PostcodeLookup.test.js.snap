// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  .c3 {
  font-size: 1rem;
  line-height: 1rem;
  text-transform: inherit;
  font-weight: bold;
  line-height: normal;
  font-family: 'Montserrat',Helvetica,Arial,sans-serif;
}

.c7 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
  padding: 0.5rem 1.25rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: 700;
  font-size: 1rem;
  border-radius: 2rem;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  height: 3.125rem;
  width: 100%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  cursor: pointer;
  background-color: #E52630;
  color: #FFFFFF;
}

.c7 > a {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c7:hover {
  background-color: #890B11;
  color: #FFFFFF;
}

.c2 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  color: #5C5C5E;
  width: 100%;
}

.c4 {
  margin-bottom: 0.5rem;
}

.c6 {
  position: relative;
  box-sizing: border-box;
  width: 100%;
  height: 48px;
  padding: 1rem 1.5rem;
  background-color: #F4F3F5;
  border: 1px solid;
  border-color: #E1E2E3;
  box-shadow: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  color: #000000;
  border-radius: 0.5rem;
  font-size: inherit;
  z-index: 2;
  font-family: 'Montserrat',Helvetica,Arial,sans-serif;
}

.c6:focus {
  border: 1px solid #666;
}

.c5 {
  position: relative;
  font-size: 1.25rem;
}

.c0 {
  position: relative;
}

.c8:disabled {
  cursor: not-allowed;
  opacity: 0.75;
}

.c9 {
  margin-top: 1.5rem;
  color: #222222;
  border: 2px solid #222222;
  background-color: #FFFFFF;
  padding-left: 3rem;
  padding-right: 3rem;
}

.c9:hover {
  color: #222222;
  background-color: #FFFFFF;
}

.c1 {
  margin-bottom: 1rem;
}

@media (min-width:740px) {
  .c7 {
    width: auto;
  }
}

@media (min-width:1024px) {
  .c7 {
    width: auto;
    padding: 1rem 2rem;
    margin: 0 auto 2rem;
  }
}

@media (min-width:740px) {
  .c6 {
    max-width: 290px;
  }
}

<div>
    <div
      className="c0 TextInputWithDropdown c1"
      onKeyDown={[Function]}
    >
      <label
        className="c2 TextInputWithDropdown__input"
        htmlFor="postcode_lookup"
      >
        <span
          className="c3 c4"
          color="inherit"
          dangerouslySetInnerHTML={
            Object {
              "__html": "Find address by postcode",
            }
          }
          size="s"
        />
        <div
          className="c5"
        >
          
          <input
            aria-describedby="postcode_lookup"
            autoComplete="off"
            className="c6"
            id="postcode_lookup"
            name="postcode_lookup"
            onChange={[Function]}
            onKeyPress={[Function]}
            placeholder="Enter postcode..."
            required={false}
            type="text"
            value=""
          />
        </div>
        
      </label>
    </div>
    <button
      className="c7 c8 c9"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Find address
      <div
        className=""
      />
    </button>
  </div>,
  .c0 {
  margin: 1rem 0 0 0.5rem;
  background: inherit;
  border: none;
  font-size: 1rem;
}

.c0:hover,
.c0:active {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

@media (min-width:740px) {

}

@media (min-width:1024px) {

}

@media (min-width:740px) {

}

<button
    className="c0"
    onClick={[Function]}
  >
    Or enter your address manually
  </button>,
]
`;
